Initial draft of tag_map user documentation (#1150)
timescale/docs: promscale/query/query-traces.md

|`is_root_span`                     | `bool`        | Is the span a root span |
|`service_name`                     | `text`        | Name of the service |
|`span_name`                        | `text`        | Name of the span |
|`span_kind`                        | `enum`        | [Span Kind](https://opentelemetry.io/docs/reference/specification/trace/api/#spankind) |
|`start_time`                       | `timestamptz` | Start date and time of the span |
|`end_time`                         | `timestamptz` | Date and time of the span end|
|`time_range`                       | `tstzrange`   | A tstzrange representation of Start and End times of the span |
|`duration_ms`                      | `float8`      | Duration of the span in milliseconds |
|`span_tags`                        | `tag_map`     | Key-value pairs for span tags. See details on `tag_map` type below |
|`dropped_tags_count`               | `int4`        | Number of dropped tags |
|`event_time`                       | `tstzrange`   | Start and end time of the event |
|`dropped_events_count`             | `int4`        | Number of dropped events |

> |`dropped_link_count`               | `int4`        | Number of dropped links |

Reviewer: Looks like this comes from the Otel Span protobuf definition, where this is called `dropped_links_count`: https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/trace/v1/trace.proto#L303

Possible values:
- `unspecified`
- `internal`
- `server`
- `client`
- `producer`
- `consumer`

### 1.1.4. `status_code` enum <a name="para-1-1-4"></a>
Possible values:
- `unset`
- `ok`
- `error`

## 1.2. Views <a name="para-1-2"></a>

To provide a stable interface to the end users a set of views is provided in the `ps_trace` schema. These include:
 - `span`
 - `link`
 - `event`

### 1.2.1. `span` view <a name="para-1-2-1"></a>

This view joins several tables to present the user with an overview of the data relevant for a single span. We've already covered its structure in 1.1. What's worth mentioning is that physically the span is stored across multiple tables and data is split across several columns for better index support. The actual table with span data is a [`timescaledb` hypertable](https://docs.timescale.com/timescaledb/latest/how-to-guides/hypertables/) which enables seamless support for retention and compression out of the box.

Below is an example of a simple query on the `span` view:

```SQL
select
        service_name,
        duration_ms
    from spans
    where
            start_time >= now() - interval'10 minutes'

>         and span_tags -> 'os_name' = '"🐧"'

Author: This was a deliberate choice to drag the attention of the reader :-)
That being said, I don't have strong feelings about this.

```SQL
select *
    from span
    where 
            start_time >= now() - interval '30 minutes'
        and span_tags -> 'pwlen' = '25'
        and resource_tags -> 'telemetry.sdk.name' = '"opentelemetry"'
```

Notice the single **and** double-quoted string `'"opentelemtry"'`. This is an artefact of PostgreSQL's `->` operator returning a `jsonb` and thus expecting a `jsonb` on the right side of the `=`. This can be worked around by using the more appropriate `->>` operator that returns text, but unfortunately that operator is not supported in the current implementation of `tag_map` type. It is planned for Promscale 1.0.0, so stay tuned!

## 3.3. Joins <a name="para-3-3"></a>

One of the main reasons to use SQL database as a backbone of your observability stack is to be able to conduct correlational analysis of different types of data, namely metrics and spans. And Promscale enables you to do exactly that.

Suppose we want to see how error rate of our service correlates with the overall memory consumption of a container. It's as easy as joining the corresponding metric table with the aforementioned `span` view:

```SQL

> select

Reviewer 2: Have you tested the query to make sure it works?

I think what would be more interesting would be to list containers and memory consumption for those containers that have a high number of errors. Then you can inspect the values and see if there is indeed a correlation.

Author: I didn't, no. It'd be great to have a place where one could test these, but the demo instance that was available during 0.5.0 was unavailable for me ever since the release.

Author: We decided to remove this query for now

## 3.4. Grouping <a name="para-3-4"></a>

One of the strengths of SQL in general and PostgreSQL in particular are aggregate functions which perform various operations on groups. We can group data on any set of columns, includinig fields of `tag_map` columns. Below is a simple example:

```SQL
select
        time_bucket('30 minutes', start_time)   as "time",
        span_tags -> 'pwlen'                    as "pwlen",
        count(*)                                as "cnt"
    from span
    where 
            start_time >= now() - interval'1 day'
        and service_name = '${service}'
    group by 1, 2
```

Here we use group by both time and the `pwlen` field of the `span_tags` column using their position in the `select` list. `count(*)` in this case is an [aggregate function](https://www.postgresql.org/docs/current/functions-aggregate.html) that counts the number of rows in the group, in our case a 30-minute window with the same value of `pwlen` field.

Now we can build a more complicated query using both techniques. Lets see how error rate correlates with the memory usage:

```SQL

> with joined as

Reviewer 2: Have you tested the query to make sure it works?

Author: Same here, we'll leave this query for a later point in time.

Below is an example of a simple query on the `span` view:

```SQL
select
        service_name,
        duration_ms
    from span
    where
            start_time >= now() - interval '10 minutes'
        and span_tags -> 'os_name' = '"linux"'
        and span_kind = 'server'
```

### 1.2.2. `event` view <a name="para-1-2-2"></a>

`event` view provides access to the events and their corresponding spans. Further details on OpenTelemetry events can be found [here](https://opentelemetry.io/docs/reference/specification/trace/api/#add-events) and [here](https://opentelemetry.io/docs/concepts/signals/traces/#span-events=).

|Name                           | Type          | Description |
|-------------------------------|---------------|-------------|
| `trace_id`                    | `trace_id`    | Trace identifier |
| `span_id`                     | `int8`        | Span identifier |

> | `time`                        | `timestamptz` | Date and time when the even has occurred |

Author: ```suggestion
| `time`                        | `timestamptz` | Date and time when the event has occurred |
```

# 2. Examples of SQL queries <a name="para-2"></a>

General notes on the queries:

> - When building timeseries graphs in Grafana, consider using the [`$__interval`](https://grafana.com/docs/grafana/latest/variables/variable-types/global-variables/#__interval) variable provided by Grafana together with TimescaleDB `time_bucket` function.

Reviewer 3: ```suggestion
* When building time series graphs in Grafana, you can use the 
   [`$__interval`](https://grafana.com/docs/grafana/latest/variables/variable-types/global-variables/#__interval) 
   variable provided by Grafana, together with the TimescaleDB `time_bucket` function.
```

> - To make the bucketing configurable, we can rely on the `timescaledb` function `time_bucket` and the corresponding Grafana variable [`$bucket_interval`](https://docs.timescale.com/timescaledb/latest/tutorials/grafana/visualizations/histograms/#prerequisites)

Reviewer 3: ```suggestion
* To make the bucketing configurable, use the TimescaleDB `time_bucket` function,
   and the corresponding Grafana variable 
  [`$bucket_interval`](https://docs.timescale.com/timescaledb/latest/tutorials/grafana/visualizations/histograms/#prerequisites)
```

> - We limit the `start_time` using grafana time filter `$__timeFilter`

Reviewer 3: ```suggestion
* Limit the `start_time` using the Grafana time filter `$__timeFilter`.
```
